### PR TITLE
op-chain-ops: artifacts FS, improve artifacts metadata

### PR DESCRIPTION
**Pull request by protolambda**
**Description**

This introduces a Go `FS` for solidity artifacts loading: a read-only filesystem interface that backs the Artifacts reads.

This way you can traverse artifacts, traverse the contracts within, and load the artifacts, all without having to think about manual filepaths.

The FS abstraction can wrap around the local filesystem (see `OpenArtifactsDir`), but could potentially also wrap around other Go FS implementations, like a tarball, or even files embedded in the Go binary. This is useful so we can archive different contracts releases, and continue to use the latest tooling around them.

This also introduces some improvements to the `Artifact` type itself:
- Support more metadata
- Handle when bytecode contains linking placeholders (we don't have linking currently, but I hit this last week when we did, which caused artifact parsing to fail)

**Tests**

Loads all the local artifacts.

**Additional context**

Part of improved op-chain-ops tooling to support running forge scripts in Go.

### CODE REVIEW

**Comment by ajsutton** [comment]
Should we start using this in https://github.com/ethereum-optimism/optimism/blob/aa501cd40a6c170008d868b5ca4c3a95ae649821/packages/contracts-bedrock/snapshots/abi_loader.go#L11 to load snapshots instead of having them directly do a fs embed? (haven't had a chance to properly review the code yet).

**Comment by protolambda** [comment]
@ajsutton perhaps. If it's more than 1 artifact, yes that would definitely make sense. This is primarily meant to navigate bundles of artifacts in a generic way. Single-use artifact files can be done whichever way.

**Comment by protolambda** [comment]
Note to self: need to update `testdata` in op-chain-ops to contain a few dummy artifacts, so the artifacts test can run against that. Currently we don't generate the contracts-bedrock artifacts before Go tests (even though we do build the contracts before the CI job runs).

**Comment by tynes** [comment]
This PR generally looks good to me, will look again once it is passing CI

**Comment by protolambda** [comment]
Ready for review

File: op-e2e/artifactsfs_test.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+		for _, contract := range contracts {
+			artifact, err := af.ReadArtifact(name, contract)
+			if err != nil {
+				if errors.Is(err, foundry.ErrLinkingUnsupported) {
+					logger.Info("linking not supported", "name", name, "contract", contract, "err", err)
+					continue
+				}
+				require.NoError(t, err, "failed to read artifact %s / %s", name, contract)
+			}
+			logger.Info("artifact",
```
Comment (tynes) [inline op-e2e/artifactsfs_test.go:34]: Do we still want to log this?

**Review by tynes — APPROVED** [review]

File: op-e2e/artifactsfs_test.go (hunk 1 above)
Comment (protolambda) [reply op-e2e/artifactsfs_test.go:34]: The log level at the top of the test is set to Warn, with a comment about it. It's there so you can see what it does if running locally. For automation we have the `NotEmpty` checks to ensure that it's looping over data.

**Pull request merged**